session/schema_amender.go

```diff
@@ -0,0 +1,556 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package session
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/pingcap/errors"
+	pb "github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/executor"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/tikv"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/rowcodec"
+	"go.uber.org/zap"
+)
+
+const amendableType = nonMemAmendType | (memBufAmendType)
+const nonMemAmendType = (1 << model.ActionAddColumn) | (1 << model.ActionDropColumn) | (1 << model.ActionDropIndex)
+const memBufAmendType = uint64(1 << model.ActionAddIndex)
+
+// Amend operation types.
+const (
+	AmendNone int = iota
+
+	// For add index.
+	AmendNeedAddDelete
+	AmendNeedAddDeleteAndInsert
+	AmendNeedAddInsert
+)
+
+// ConstOpAddIndex is the possible ddl state changes, and related amend action types.
+var ConstOpAddIndex = map[model.SchemaState]map[model.SchemaState]int{
+	model.StateNone: {
+		model.StateDeleteOnly:          AmendNeedAddDelete,
+		model.StateWriteOnly:           AmendNeedAddDeleteAndInsert,
+		model.StateWriteReorganization: AmendNeedAddDeleteAndInsert,
+		model.StatePublic:              AmendNeedAddDeleteAndInsert,
+	},
+	model.StateDeleteOnly: {
+		model.StateWriteOnly:           AmendNeedAddInsert,
+		model.StateWriteReorganization: AmendNeedAddInsert,
+		model.StatePublic:              AmendNeedAddInsert,
+	},
+	model.StateWriteOnly: {
+		model.StateWriteReorganization: AmendNone,
+		model.StatePublic:              AmendNone,
+	},
+	model.StateWriteReorganization: {
+		model.StatePublic: AmendNone,
+	},
+}
+
+type schemaAndDecoder struct {
+	schema  *expression.Schema
+	decoder *rowcodec.ChunkDecoder
+}
+
+// amendCollector collects all amend operations, row decoders and memory chunks for each table needs amend.
+type amendCollector struct {
+	tblAmendOpMap map[int64][]amendOp
+	tblDecoder    map[int64]*schemaAndDecoder
```

Comment:
I think we don't need to maintain `tblDecoder` and `tblChk` in `amendCollector`.
We can create them directly when creating amendIndexInfo.

Reply:
Do you mean creating a new one for each `amendOperationAddIndexInfo` and it's not needed to reuse?

Reply:
Yes, in most cases, we only have one index to fix.